### ISSUE TITLE
`x86_64-darwin` -> `wasm32-unknown-unknown` strip occasionally fails

### ISSUE THREAD

**Issue by rvolosatovs** [ipetkov/crane#453]
That's an interesting one - building this https://github.com/rvolosatovs/nixify/tree/a181712ef4456ea1a671e3515e0f8d4925557161/examples/rust-complex cargo workspace from `x86_64-darwin` towards `wasm32-unknown-unknown` started failling after update from v0.13.1 to v0.15.0 with:
```
rust-complex> strip: warning: can't process non-object and non-archive file: /nix/store/58n6yxijpyb5dfrf0pcjakjhj72lk8cr-rust-complex-0.0.0-unspecified/bin/foo.wasm
rust-complex> strip: warning: can't process non-object and non-archive file: /nix/store/58n6yxijpyb5dfrf0pcjakjhj72lk8cr-rust-complex-0.0.0-unspecified/lib/foo.wasm
rust-complex> xargs: strip: terminated by signal 11
```

Builds from `x86_64-linux`, `aarch64-linux` and `aarch64-darwin` are fine.

What's more, building a simple hello-world project (https://github.com/rvolosatovs/nixify/tree/a181712ef4456ea1a671e3515e0f8d4925557161/examples/rust-hello) also works just fine

It appears to be a regression introduced somewhere along the way from `v0.13.1` to `v0.15.0`: https://github.com/ipetkov/crane/compare/v0.13.1...v0.15.0 and very likely related to `cdylib`/`staticlib` crate types. I don't have the resources right now to produce a minimized reproducer, but decided to open an issue to point this out regardless.

**Comment by ipetkov** [comment]
Hi @rvolosatovs thanks for the report! Do you have a `flake.lock` which can reproduce the issue?

I ran the following on an x86 Darwin machine, each succeeded on the first try _and could not reproduce the error_:
* `nix build -L 'github:rvolosatovs/nixify/a181712ef4456ea1a671e3515e0f8d4925557161/?dir=examples/rust-complex#default' --no-write-lock-file`
* `nix flake check -L 'github:rvolosatovs/nixify/a181712ef4456ea1a671e3515e0f8d4925557161/?dir=examples/rust-complex' --no-write-lock-file`

**Comment by rvolosatovs** [comment]
> Hi @rvolosatovs thanks for the report! Do you have a `flake.lock` which can reproduce the issue?
> 
> I ran the following on an x86 Darwin machine, each succeeded on the first try _and could not reproduce the error_:
> 
>     * `nix build -L 'github:rvolosatovs/nixify/a181712ef4456ea1a671e3515e0f8d4925557161/?dir=examples/rust-complex#default' --no-write-lock-file`
> 
>     * `nix flake check -L 'github:rvolosatovs/nixify/a181712ef4456ea1a671e3515e0f8d4925557161/?dir=examples/rust-complex' --no-write-lock-file`

Thanks for the prompt response.

https://github.com/rvolosatovs/nixify/actions/runs/6876729059/job/18702934233 here's a link to CI failure. I was able to reproduce it locally, but, weirdly enough, I cannot do so anymore.

Neither by

```
nix build -L --no-write-lock-file "github:rvolosatovs/nixify/e6aa0aec7e9cf7d1deb64a611b0a6e8b953b082b?dir=examples/rust-complex#packages.x86_64-darwin.rust-complex-debug-wasm32-unknown-unknown"
```

nor checking out the repo at e6aa0aec7e9cf7d1deb64a611b0a6e8b953b082b manually and doing:

```
nix build --override-input nixify ./. -L "./examples/rust-complex#packages.x86_64-darwin.rust-complex-debug-wasm32-unknown-unknown"
```

Unfortunately, it appears that Nix evaluations on Darwin are non-deterministic - I don't know if it has something to do with https://github.com/DeterminateSystems/nix-installer or a general Nix-on-Darwin bug, but it's not the first time I encounter non-determinism running Nix on Darwin in the last months.

In fact, running `nix flake check -L github:rvolosatovs/nixify` on aarch64-darwin is almost guaranteed to trigger (what I so far believe to be) a race condition at least on my system, which manifests as `cargo` failing build with "missing target", which does not happen with the exact same code path on Linux.
If you're interested, the relevant code "constructing the toolchain" is here https://github.com/rvolosatovs/nixify/blob/262c43a645400e68ea63eaede81b18281b846b93/lib/rust/mkAttrs.nix#L247-L269 - by inspecting the resulting failing derivation, the Rust toolchain input points to the correct toolchain (i.e. it has the targets installed and is able to build for, for example, `wasm32-wasi`), but the build fails regardless because *somehow* the wrong toolchain ends up being used in the derivation *sometimes* (e.g. it may work after clearing the cached build from store)

Closing this as this does not look like `crane` issue, hopefully I will be able to produce a minimize reproducer for the non-determinism I'm seeing to file an issue upstream (in https://github.com/NixOS/nix, I suppose) eventually.
